# Activity 3: Practicing with Pandas

For this activity, we will assume the role of a data scientist in the search for and evaluation of public datasets. In turn, pay close attention to the dataset you choose for counter narratives or nonapparent political values, gaps or underrepresented perspectives.

Please make sure you open this file in Jupyter Notebook or Google Colab -- you will have to either open Jupyter Notebook first or upload the file to Google Colab before being able to edit the file.

Please make sure you **print your output.** You may need to use ```print``` for your output to be displayed.

-----
## Submission Instructions
1. Upload the completed activity to the "Activities" folder of your GitHub Portfolio;
2. Ensure your submission consists of two files: <br/>(a) one single-page reflection in markdown (e.g. `a3_reflection.md`);  <br/>(b) one Jupyter notebook with completed responses to each activity step (e.g. `a3_pandas.ipynb`).

## Data biography: Trans-Atlantic Slave Trade dataset

To complete the assignment, please make sure you download the dataset from where it hosted on our GitHub: [Trans-Atlantic-Slave-Trade_Americas.csv](https://github.com/zmuhls/ccny-data-science/blob/main/assets/datasets/Trans-Atlantic-Slave-Trade_Americas.csv)

## Data Biography (1-2 Paragraphs)

Before you begin your exploration, you will provide a short data biography (1 - 2 paragraphs) on the dataset that addresses **where the data came from, who collected it, and the original intention(s) for its collection.**

You may find the following sources useful for your research:
* [Slave Voyages' Trans-Atlantic Slave Trade methodology](https://www.slavevoyages.org/voyage/about#methodology/introduction/0/en/)
    * You may find the sections: Introduction, Coverage of the Slave Trade, and Nature of Sources particularly useful.
* [Slave Voyages' About](https://www.slavevoyages.org/about/about#)
* [Jamelle Bouie's We Still Can’t See American Slavery for What It Was](https://csc10800.github.io/assets/pdf/Bouie_we_still_cant_see_american_slavery_for_what_it_is.pdf)

**Enclose Data Biography below (1-2 Paragraph)

*   List item
*   List item

**:


-----

## Download and explore dataset

As we have been practicing in class, the questions that follow will require you to consider what output you will need before running the appropriate python codes.

To ensure your code runs, please remember to load ```pandas``` and the ```Trans-Atlantic-Slave-Trade_Americas.csv``` file into the jupyter notebook environment.

If you are running this on Google Colab, please make sure you connect your google drive with your Colab notebook first before attempting the exercise:
```
from google.colab import drive
drive.mount('/content/drive')
```

In [ ]:
# Your code here

### 2. How many rows of data does the dataset contain?</br>
In this first step, you’ll load the dataset and examine its structure. Loading data is foundational in data science and digital humanities workflows. You’ll use the `pandas` library to read a CSV file and display its first few rows.</br>
* **Hint**: See Melanie Walsh's chapter on Pandas Basics if you're stuck: [Pandas Basics – Part 1](https://melaniewalsh.github.io/Intro-Cultural-Analytics/03-Data-Analysis/01-Pandas-Basics-Part1.html)).



In [2]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/zmuhls/ccny-data-science/main/assets/datasets/Trans-Atlantic-Slave-Trade_Americas.csv"
data = pd.read_csv(url)

# Get the number of rows
num_rows = data.shape[0]

# Print the result
print("Number of rows in the dataset:", num_rows)


Number of rows in the dataset: 20779


**Number of rows in the dataset: 20779**     

</br>
</br>
</br>

---


### 3. Are the data types for each column appropriate? Please explain how they are/not approrpiate for your analysis.

The data types for each column are mostly fine, but some could be adjusted depending on what we need to do with them. For example:

~If there are any numeric columns stored as strings (like population or numbers), they’d need to be converted to integers or floats to do calculations.
~Date columns, if they exist, should be in a proper datetime format so we can easily filter or analyze them by time.
Categorical data, like regions or countries, might be better stored as a "category" type to save memory and make processing faster.
~Overall, the data types might work as is for basic analysis, but we’d need to check them based on what kind of analysis we’re doing.

### 4. What is the overall average proportion of ```percent_women```, ```percent_children```, ```percent_men```?

In [4]:
average_percent_women = data['percent_women'].mean()
average_percent_children = data['percent_children'].mean()
average_percent_men = data['percent_men'].mean()

# Print the results
print("Average Proportion of Women:", average_percent_women)
print("Average Proportion of Children:", average_percent_children)
print("Average Proportion of Men:", average_percent_men)


Average Proportion of Women: 0.27409773997457043
Average Proportion of Children: 0.23153138034413165
Average Proportion of Men: 0.4970504753391753


**Double click to edit cell**     
`Please type your answer here`
</br>
</br>
</br>


---


### 5. How many of the column values for ```percent_women```, ```percent_children```, ```percent_men``` are left blank? Suggest 1 reason why majority of the values in these columns are blank.

In [5]:
# Count the number of missing (blank) values in each column
missing_percent_women = data['percent_women'].isnull().sum()
missing_percent_children = data['percent_children'].isnull().sum()
missing_percent_men = data['percent_men'].isnull().sum()

# Print the results
print("Number of blank values in percent_women:", missing_percent_women)
print("Number of blank values in percent_children:", missing_percent_children)
print("Number of blank values in percent_men:", missing_percent_men)


Number of blank values in percent_women: 17869
Number of blank values in percent_children: 17831
Number of blank values in percent_men: 17869


**A lot of these are probably blank because they didn’t bother recording the exact breakdowns back then. They were more focused on overall numbers and might not have cared enough about the details, or the info just wasn’t available. Record-keeping wasn’t super reliable in that era.**



</br>
</br>
</br>

---


### 6. Display all duplicated rows and remove the duplicates. Check that the duplicates were successfully removed. Please also suggest a reason why we would remove duplicates for our analysis.

**Level-up**: How many duplicated rows do we have? Recall the methods we have used to help us count things.

In [6]:
import pandas as pd

duplicated_rows = data[data.duplicated()]
print("Duplicated rows:")
print(duplicated_rows)

num_duplicates = data.duplicated().sum()
print("\nNumber of duplicated rows:", num_duplicates)

data_cleaned = data.drop_duplicates()

num_duplicates_after = data_cleaned.duplicated().sum()
print("\nNumber of duplicated rows after cleaning:", num_duplicates_after)


Duplicated rows:
       year_of_arrival              flag  \
213               1577               NaN   
214               1577               NaN   
215               1577               NaN   
283               1586               NaN   
285               1586               NaN   
286               1586               NaN   
769               1602               NaN   
770               1602               NaN   
1665              1654     Great Britain   
1666              1654     Great Britain   
1667              1654     Great Britain   
9842              1763       Netherlands   
9843              1763       Netherlands   
9844              1763       Netherlands   
9845              1763       Netherlands   
10356             1765            France   
10357             1765            France   
10358             1765            France   
10880             1767            France   
10881             1767            France   
10882             1767            France   
14574          

Duplicates can mess up your analysis by double-counting certain records. For example, if the same voyage appears twice, it can make averages, totals, or proportions look incorrect. Removing duplicates ensures your analysis reflects accurate, unique data.
</br>


---


### 7. Please identify the **top 5 most common ports of arrival**.

**Hint:** Check the columns ```place_of_landing```.

In [8]:
# Step 1: Identify the top 5 most common ports of arrival
top_ports = data['place_of_landing'].value_counts().head(5)

# Step 2: Print the results
print("Top 5 most common ports of arrival:")
print(top_ports)

Top 5 most common ports of arrival:
place_of_landing
Barbados, port unspecified    2041
Jamaica, port unspecified     1715
Kingston                      1627
Havana                        1327
Cap Francais                  1127
Name: count, dtype: int64


**Double click to edit cell**     
`Please type your answer here`
</br>
</br>
</br>

---


# 8. Please plot the top 5 enslavers/captors from this dataset. Please choose the appropriate visualization (e.g. pie chart, bar chart) to display your finding.



In [17]:
import pandas as pd
import matplotlib.pyplot as plt

# Step 1: Load the dataset
url = "https://raw.githubusercontent.com/zmuhls/ccny-data-science/main/assets/datasets/Trans-Atlantic-Slave-Trade_Americas.csv"
data = pd.read_csv(url)

# Step 2: Check available columns
print("Available columns in the dataset:")
print(data.columns)

# Step 3: Inspect the first few rows to understand the data
print("\nPreview of the dataset:")
print(data.head())

# Step 4: Replace 'CORRECT_COLUMN_NAME' with the actual column name after inspecting the data
# Uncomment and update the code below once you identify the correct column name
# top_enslavers = data['CORRECT_COLUMN_NAME'].value_counts().head(5)

# Uncomment to plot the data once the column name is identified
# plt.figure(figsize=(8, 5))
# top_enslavers.plot(kind='bar', color='skyblue')
# plt.title("Top 5 Enslavers/Captors", fontsize=16)
# plt.xlabel("Enslavers/Captors", fontsize=12)
# plt.ylabel("Count", fontsize=12)
# plt.xticks(rotation=45)
# plt.tight_layout()
# plt.show()


Available columns in the dataset:
Index(['year_of_arrival', 'flag', 'place_of_purchase', 'place_of_landing',
       'percent_women', 'percent_children', 'percent_men', 'total_embarked',
       'total_disembarked', 'resistance_label', 'vessel_name',
       'captain's_name', 'voyage_id', 'sources'],
      dtype='object')

Preview of the dataset:
   year_of_arrival               flag   place_of_purchase  \
0             1520                NaN   Portuguese Guinea   
1             1525  Portugal / Brazil            Sao Tome   
2             1526    Spain / Uruguay  Cape Verde Islands   
3             1526    Spain / Uruguay  Cape Verde Islands   
4             1526                NaN  Cape Verde Islands   

                 place_of_landing  percent_women  percent_children  \
0                        San Juan            NaN               NaN   
1         Hispaniola, unspecified            NaN               NaN   
2          Cuba, port unspecified            NaN               NaN   
3      

-----

### 9. Having briefly explored the dataset, what further questions have emerged as you explored this dataset?    

Please share your thoughts in a few sentences.


a few questions came to mind...


*   Are there patterns or trends in the demographics (percentages of men, women, and children) across different ports of landing or over time?




-----
-----


## Side quest challenge

The side quest challenge is for extra credit. You can still get full credit for this activity even if you do not complete this challenge.

### 1. Which enslaver/captor had the highest difference between the total number of people who embarked and disembarked?

**Hint:** You may need to [add a column](https://github.com/GCDigitalFellows/intro-pandas-dri-2022/blob/main/README.md#8-rename-select-drop-and-add-new-columns) to calculate the difference, use the [```.groupby```, ```.count()``` and ```.sort_values``` methods](https://github.com/GCDigitalFellows/intro-pandas-dri-2022/blob/main/README.md#9-sort-columns-groupby-columns--count-values) for this challenge.

In [2]:
import pandas as pd

# Step 1: Load the dataset
url = "https://raw.githubusercontent.com/zmuhls/ccny-data-science/main/assets/datasets/Trans-Atlantic-Slave-Trade_Americas.csv"
data = pd.read_csv(url)

# Step 2: Add a new column to calculate the difference between embarked and disembarked
data['difference'] = data['total_embarked'] - data['total_disembarked']

# Step 3: Group by the appropriate column (likely 'captain\'s_name')
# Replace 'captain\'s_name' with another grouping column if needed
grouped_data = data.groupby("captain's_name")['difference'].sum()

# Step 4: Sort the results to find the enslaver/captor with the highest difference
sorted_data = grouped_data.sort_values(ascending=False)

# Step 5: Display the top result
print("Enslaver/Captor with the highest difference:")
print(sorted_data.head(1))


Enslaver/Captor with the highest difference:
captain's_name
Smith, John    1143.0
Name: difference, dtype: float64


**Double click to edit cell**     
`Please type your answer here`
</br>
</br>
</br>

---
